test/extended/dr/machine_recover.go

```diff
@@ -0,0 +1,276 @@
+package dr
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"go.etcd.io/etcd/pkg/transport"
+	"google.golang.org/grpc"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/openshift/origin/test/extended/util/disruption"
+	"github.com/openshift/origin/test/extended/util/disruption/controlplane"
+	"github.com/openshift/origin/test/extended/util/disruption/frontends"
+)
+
+var _ = g.Describe("[Feature:DisasterRecovery][Disruptive]", func() {
+	f := framework.NewDefaultFramework("machine-recovery")
+	f.SkipNamespaceCreation = true
+	f.SkipPrivilegedPSPBinding = true
+
+	oc := exutil.NewCLIWithoutNamespace("machine-recovery")
+
+	g.It("[dr-quorum-restore] Cluster should survive master and worker failure and recover with machine health checks", func() {
+		framework.SkipUnlessProviderIs("aws")
+
+		disruption.Run("Machine Shutdown and Restore", "machine_failure",
+			disruption.TestData{},
+			[]upgrades.Test{
+				&upgrades.ServiceUpgradeTest{},
+				&controlplane.AvailableTest{},
+				&frontends.AvailableTest{},
+			},
+			func() {
+
+				config, err := framework.LoadConfig()
+				o.Expect(err).NotTo(o.HaveOccurred())
+				dynamicClient := dynamic.NewForConfigOrDie(config)
+				ms := dynamicClient.Resource(schema.GroupVersionResource{
+					Group:    "machine.openshift.io",
+					Version:  "v1beta1",
+					Resource: "machines",
+				}).Namespace("openshift-machine-api")
+
+				framework.Logf("Verify SSH is available before restart")
+				masters, workers := clusterNodes(oc)
+				o.Expect(len(masters)).To(o.BeNumerically(">=", 3))
+				o.Expect(len(workers)).To(o.BeNumerically(">=", 2))
+
+				framework.Logf("Verify etcd endpoints are healthy")
+				certsDir := "/tmp/etcd-client-certs/"
+				dumpEtcdCertsOnDisk(oc, certsDir)
+				defer os.RemoveAll(certsDir)
+				healthyEtcds := getHealthyEtcds(certsDir, masters)
+				o.Expect(len(healthyEtcds)).To(o.BeNumerically("==", len(masters)))
```

Comment:
Might be better to gate this on 3 (we expect a quorum before we start) rather than number of masters, since number of masters might be 4.

Reply:
No unhealthy etcd, must have at least 3, or something.

Reply:
good idea will modify to check:
1. 0 unhealthy before starting
2. 1 unhealthy after the master is destroyed
3. 0 unhealthy after master recreated

Reply:
modified